File: src/main/java/data/stack/ExclusiveTimeOfFunctions_636.java
package data.stack;

import java.util.*;

public class ExclusiveTimeOfFunctions_636 {
  public static int[] exclusiveTime(int n, List<String> logs) {
    var executions = new ArrayList<>(logs.stream().map(l -> {
      var arr = l.split(":");
      return new Log(Integer.parseInt(arr[0]), arr[1].equals("start"), Integer.parseInt(arr[2]));
    }).toList());

    executions.sort(Comparator.comparingInt(e -> e.time));

    var times = new int[n];
    var stack = new Stack<Log>();
    for (var e : executions) {
      if (e.start) {
        stack.push(e);
      } else {
        var log = stack.pop();
        var time = e.time - log.time + 1;
        times[log.id] += time;
        if (!stack.isEmpty()) {
          times[stack.peek().id] -= time;
        }
      }
    }

    return times;
  }

  private record Log(int id, boolean start, int time) {}

  public static void main(String[] args) {
//    var result = exclusiveTime(2, List.of("0:start:0","1:start:2","1:end:5","0:end:6"));
//    var result = exclusiveTime(1, List.of("0:start:0","0:start:2","0:end:5","0:start:6","0:end:6","0:end:7"));
//    var result = exclusiveTime(2, List.of("0:start:0","0:start:2","0:end:5","1:start:6","1:end:6","0:end:7"));
//    var result = exclusiveTime(2, List.of("0:start:0","0:start:2","0:end:5","1:start:7","1:end:7","0:end:8"));
//    var result = exclusiveTime(1, List.of("0:start:0","0:start:1","0:start:2","0:end:3","0:end:4","0:end:5"));
    var result = exclusiveTime(3, List.of("1:start:0","0:start:2","1:start:3","2:start:4","2:end:4","0:end:6","1:end:7","1:end:8"));
    Arrays.stream(result).asLongStream().forEach(System.out::println);
  }
}

File: src/main/kotlin/data/structure/stack/ExclusiveTimeOfFunctions_636.kt
package data.structure.stack

import java.util.Stack

class KtExclusiveTimeOfFunctions_636 {
    fun exclusiveTime(n: Int, logs: List<String>): IntArray {
        val exclusiveTime = IntArray(n)
        val stack = Stack<FunctionLog>()
        logs.map { log ->
            val fields = log.split(":")
            val start = fields[1] == "start"
            FunctionLog(fields[0].toInt(), start, fields[2].toInt())
        }.sortedBy { it.timestamp }.forEach { log ->
            if (log.start) {
                stack.push(log)
            } else {
                val prev = stack.pop()
                val time = log.timestamp - prev.timestamp + 1
                // time for prev function
                exclusiveTime[prev.id] += time

                // if previous function is not finished, we need to subtract current time for previous one
                if (stack.isNotEmpty() && stack.peek().start) {
                    exclusiveTime[stack.peek().id] -= time
                }
            }
        }

        return exclusiveTime
    }

    data class FunctionLog(val id: Int, val start: Boolean, val timestamp: Int)
}


fun main() {
    val time = KtExclusiveTimeOfFunctions_636()
    println(time.exclusiveTime(2, listOf("0:start:0","1:start:2","1:end:5","0:end:6")).asList())
    println(time.exclusiveTime(3, listOf("1:start:0","0:start:2","1:start:3","2:start:4","2:end:4","0:end:6","1:end:7","1:end:8")).asList())
}